external/js/cothority/src/network/connection.ts

```diff
@@ -242,3 +250,234 @@ export class LeaderConnection extends WebSocketConnection {
+     * with a node from the reserve queue.
+     * @param address node with successful return
+     * @param rang order of arrival
+     * @param ratio delay in answer between first reply and this reply
+     */
+    done(address: string, rang: number, ratio: number) {
+        const index = this.index(address);
+        if (index >= 0) {
+            if (ratio >= Nodes.slowThreshold) {
+                this.replaceActive(index);
+            } else {
+                this.swapActive(index, rang);
+            }
+        }
+    }
+
+    /**
+     * Sets the timeout on all nodes.
+     * @param t
+     */
+    setTimeout(t: number) {
+        this.active.concat(this.reserve).forEach((n) => n.setTimeout(t));
+    }
+
+    /**
+     * Replaces the given node from the active queue with the first node from
+     * the reserve queue.
+     * @param index
+     */
+    private replaceActive(index: number) {
+        if (index >= 0) {
+            this.reserve.push(this.active.splice(index, 1)[0]);
+            this.active.push(this.reserve.shift());
+        }
+    }
+
+    /**
+     * Swaps two nodes in the active queue.
+     */
+    private swapActive(a: number, b: number) {
+        if (a >= 0 && b >= 0 &&
+            a < this.active.length && b < this.active.length) {
+            [this.active[a], this.active[b]] =
+                [this.active[b], this.active[a]];
+        } else {
+            Log.error("asked to swap", a, b, this.active.length);
+        }
+    }
+
+    /**
+     * Gets the index of a given address.
+     * @param address
+     */
+    private index(address: string): number {
+        return this.active.findIndex((c) => {
+            return c.address === address;
+        });
+    }
+}
+
+/**
+ * A Node holds one WebSocketConnection per service.
+ */
+export class Node {
+    private services: Map<string, WebSocketConnection> = new Map<string, WebSocketConnection>();
+
+    constructor(readonly address: string) {
+    }
+
+    /**
+     * Returns a WebSocketConnection for a given service. If the
+     * connection doesn't exist yet, it will be created.
+     * @param name
+     */
+    getService(name: string): WebSocketConnection {
+        if (this.services.has(name)) {
+            return this.services.get(name);
+        }
+        this.services.set(name, new WebSocketConnection(this.address, name));
+        return this.getService(name);
+    }
+
+    /**
+     * Sets the timeout for all connections of this node.
+     */
+    setTimeout(t: number) {
+        this.services.forEach((conn) => conn.setTimeout(t));
+    }
+}
+
+/**
+ * A NodeList is used to interact with the Nodes-class by allowing
+ * the requester to indicate the order of arrival of messages and
+ * which nodes didn't reply correctly.
+ */
+export class NodeList {
+    readonly active: WebSocketConnection[];
+    private reserve: WebSocketConnection[];
+    private readonly start: number;
+    private first: number;
```

Comment:
```suggestion
    private first: number = 0;
```

Reply:
that's not by default in JS?

Reply:
Not in JS at least, it just a name without a type, and I don't think TypeScript does it.